File: competitive-events-api/src/main/java/es/jocasolo/competitiveeventsapi/service/PunishmentService.java
package es.jocasolo.competitiveeventsapi.service;

import org.springframework.web.multipart.MultipartFile;

import es.jocasolo.competitiveeventsapi.dto.punishment.PunishmentDTO;
import es.jocasolo.competitiveeventsapi.dto.punishment.PunishmentPostDTO;
import es.jocasolo.competitiveeventsapi.dto.punishment.PunishmentPutDTO;
import es.jocasolo.competitiveeventsapi.exceptions.event.EventNotFoundException;
import es.jocasolo.competitiveeventsapi.exceptions.image.ImageUploadException;
import es.jocasolo.competitiveeventsapi.exceptions.punishment.PunishmentNotFoundException;
import es.jocasolo.competitiveeventsapi.exceptions.punishment.PunishmentWrongUpdateException;
import es.jocasolo.competitiveeventsapi.exceptions.user.UserNotValidException;
import es.jocasolo.competitiveeventsapi.model.event.Punishment;

public interface PunishmentService {

	/**
	 * @param id
	 * @return
	 * @throws PunishmentNotFoundException
	 */
	Punishment findOne(Integer id) throws PunishmentNotFoundException;

	/**
	 * @param userDto
	 * @return
	 * @throws EventNotFoundException 
	 * @throws UserNotValidException 
	 */
	PunishmentDTO create(PunishmentPostDTO punishmentDto) throws EventNotFoundException, UserNotValidException;

	/**
	 * @param id
	 * @param punishmentDto
	 * @throws PunishmentNotFoundException 
	 * @throws UserNotValidException 
	 * @throws PunishmentWrongUpdateException 
	 */
	void update(Integer id, PunishmentPutDTO punishmentDto) throws PunishmentNotFoundException, UserNotValidException, PunishmentWrongUpdateException;

	/**
	 * @param id
	 * @throws EventNotFoundException 
	 * @throws UserNotValidException 
	 * @throws PunishmentNotFoundException 
	 */
	void delete(Integer id) throws EventNotFoundException, UserNotValidException, PunishmentNotFoundException;

	/**
	 * @param id
	 * @param file
	 * @return
	 * @throws ImageUploadException 
	 * @throws PunishmentNotFoundException 
	 * @throws UserNotValidException 
	 */
	Object updateImage(Integer id, MultipartFile file) throws ImageUploadException, PunishmentNotFoundException, UserNotValidException;
	
}

File: competitive-events-app/app/src/main/java/es/jocasolo/competitiveeventsapp/service/PunishmentService.kt
package es.jocasolo.competitiveeventsapp.service

import es.jocasolo.competitiveeventsapp.dto.punishment.PunishmentDTO
import es.jocasolo.competitiveeventsapp.dto.punishment.PunishmentPostDTO
import okhttp3.MultipartBody
import retrofit2.Call
import retrofit2.http.*

interface PunishmentService {

    @Headers("Content-Type: application/json")
    @GET("/punishments/{id}")
    fun findOne(@Path (value="id") id: String, @Header(value="Authorization") authorization : String) : Call<PunishmentDTO>

    @Headers("Content-Type: application/json")
    @POST("/punishments")
    fun create(@Body body: PunishmentPostDTO, @Header(value="Authorization") authorization : String) : Call<PunishmentDTO>

    @Headers("Content-Type: application/json")
    @DELETE("/punishments/{id}")
    fun delete(@Path (value="id") id: Int, @Header(value="Authorization") authorization : String) : Call<Void>

    @Multipart
    @PUT("/punishments/{id}/image")
    fun updateImage(@Part file : MultipartBody.Part, @Path (value="id") id: Int, @Header(value="Authorization") authorization : String) : Call<PunishmentDTO>

}